File: src/main/java/hzt/aoc/day25/Part1ComboBreaker.java
package hzt.aoc.day25;

public class Part1ComboBreaker extends Day25Challenge {

    private static final long STARTING_VALUE = 1;

    public Part1ComboBreaker() {
        super("part 1",
                "What encryption key is the handshake trying to establish?");
    }

    @Override
    protected long solveByInput(final long cardPublicKey, final long doorPublicKey) {
        final long loopSizeCard = findLoopSize(cardPublicKey);
        final long loopSizeDoor = findLoopSize(doorPublicKey);
        final long encryptionKeyDoor = calculateEncryptionKey(doorPublicKey, loopSizeCard);
        final long encryptionKeyCard = calculateEncryptionKey(cardPublicKey, loopSizeDoor);
        LOGGER.trace("Card loop size: " + loopSizeCard);
        LOGGER.trace("Door loop size: " + loopSizeDoor);
        LOGGER.trace("Card encryption key: " + encryptionKeyCard);
        LOGGER.trace("Door encryption key: " + encryptionKeyDoor);
        return encryptionKeyCard == encryptionKeyDoor ? encryptionKeyCard : 0;
    }

    private long calculateEncryptionKey(final long publicKey, final long loopSizeOther) {
        long value = STARTING_VALUE;
        for (int i = 0; i < loopSizeOther; i++) {
            value = performStep(value, publicKey);
        }
        return value;
    }

    private long performStep(long value, final long subjectNumber) {
        value = value * subjectNumber;
        value = value % NUMBER_TO_DIVIDE_BY;
        return value;
    }

    private long findLoopSize(final long publicKey) {
        long newVal = STARTING_VALUE;
        int counter = 0;
        do {
            newVal = performStep(newVal, INIT_SUBJECT_NUMBER);
            counter++;
        } while (newVal != publicKey);
        return counter;
    }

    @Override
    String getMessage(final long global) {
        return String.format("%d", global);
    }

}

File: src/main/kotlin/hzt/aoc/day25/Part1ComboBreaker.kt
package hzt.aoc.day25

class Part1ComboBreaker : Day25Challenge(
    "part 1",
    "What encryption key is the handshake trying to establish"
) {
    override fun solveByInput(cardPublicKey: Long, doorPublicKey: Long): Long {
        val loopSizeCard = findLoopSize(cardPublicKey)
        val loopSizeDoor = findLoopSize(doorPublicKey)
        val encryptionKeyDoor = calculateEncryptionKey(doorPublicKey, loopSizeCard)
        val encryptionKeyCard = calculateEncryptionKey(cardPublicKey, loopSizeDoor)
        LOGGER.trace("Card loop size: $loopSizeCard")
        LOGGER.trace("Door loop size: $loopSizeDoor")
        LOGGER.trace("Card encryption key: $encryptionKeyCard")
        LOGGER.trace("Door encryption key: $encryptionKeyDoor")
        return if (encryptionKeyCard == encryptionKeyDoor) encryptionKeyCard else 0
    }

    private fun calculateEncryptionKey(publicKey: Long, loopSizeOther: Long): Long {
        var value = STARTING_VALUE
        for (i in 0 until loopSizeOther) {
            value = performStep(value, publicKey)
        }
        return value
    }

    private fun performStep(value: Long, subjectNumber: Long): Long = (value * subjectNumber) % NUMBER_TO_DIVIDE_BY

    private fun findLoopSize(publicKey: Long): Long {
        var newVal = STARTING_VALUE
        var counter = 0
        do {
            newVal = performStep(newVal, INIT_SUBJECT_NUMBER.toLong())
            counter++
        } while (newVal != publicKey)
        return counter.toLong()
    }

    override fun getMessage(value: Long): String {
        return String.format("%d", value)
    }

    companion object {
        private const val STARTING_VALUE: Long = 1
    }
}
